contracts/competition/Competition.sol

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+import "../referrals/interfaces/IReferralStorage.sol";
+import "../access/Governable.sol";
+
+contract Competition is Governable
+{
+    struct Team {
+        address leader;
+        string name;
+        bytes32 referral;
+        address[] members;
+        address[] joinRequests;
+    }
+
+    struct CompetitionDetails {
+        uint start;
+        uint end;
+        uint maxTeamSize;
+    }
+
+    CompetitionDetails public competitionDetails;
+    IReferralStorage public referralStorage;
+    address[] public leaders;
+    mapping(address => Team) public teams;
+    mapping(string => bool) public teamNames;
+    mapping(address => address) public membersToTeam;
+    mapping(address => address) public requests;
+
+    event TeamRegistered(address leader, string name, bytes32 referral);
+    event JoinRequestCreated(address member, address leader);
+    event JoinRequestCanceled(address member);
+    event JoinRequestApproved(address member, address leader);
+    event MemberRemoved(address leader, address member);
+    event CompetitionDetailsChanged(uint start, uint end, uint maxTeamSize);
+
+    modifier registrationIsOpen() {
+        require(block.timestamp < competitionDetails.start, "Registration is closed.");
+        _;
+    }
+
+    modifier isNotMember() {
+        require(membersToTeam[msg.sender] == address(0), "Team members are not allowed.");
+        _;
+    }
+
+    constructor(
+        uint start,
+        uint end,
+        uint maxTeamSize,
+        IReferralStorage referralStorage_
```

Comment:
in other contracts to avoid variable conflicts we add leading underscore to func args instead of trailing slash. i think it's good to have it consistent